src/ui/base-page-storyEdit.js

```diff
@@ -334,6 +447,43 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/main/LICENSE.
         }
     };
 
+    /**
+     * Increments or decrements a pair of counters depending on the previous and new
+     * states of a particular value. The expected possible state transitions are:
+     *
+     * - undefined to ready
+     * - ready to uploading
+     * - uploading to ready
+     * - uploading to errorReceived
+     * - errorReceived to uploading
+     *
+     * @typedef {Object.<String, Number>} UploadStateCounters
+     * @property {Number} UploadStateCounters.errors - the number of blocks in an "errorReceived" state
+     * @property {Number} UploadStateCounters.uploads - the number of blocks in an "uploading" state
+     *
+     * @param {String} previousState - the previous state of the value
+     * @param {String} newState - the new state of the value
+     * @param {Component} storyEdit - an instance of `sjrk.storyTelling.base.page.storyEdit`
+     */
+    sjrk.storyTelling.base.page.storyEdit.updateUploadCounters = function (previousState, newState, storyEdit) {
+        var counters = fluid.copy(storyEdit.model.uploadCounters);
+
+        if (previousState === "errorReceived") {
+            counters.errors--;
+            counters.uploads++;
+        } else if (previousState === "uploading") {
+            counters.uploads--;
+
+            if (newState === "errorReceived") {
+                counters.errors++;
+            }
+        } else if (previousState === "ready") {
+            counters.uploads++;
+        }
+
+        storyEdit.applier.change(["uploadCounters"], counters);
```

Comment:
Unless this is handled in a model relay, it doesn't appear that there are any limits on the range. At least we shouldn't allow the count to go below 0.

Reply:
I don't think it's possible for the counters to go below 0, as neither of them is decremented when the `previousState` is not one of the three cases handled by the outer `if` block. Assuming model relays & the `changeApplier` are synchronous, i.e. they won't be able to run simultaneously, then the same counter could not be decremented without a corresponding initial increment. I can add the following just before the `changeApplier` call, if you think it would be helpful:
```JavaScript
// ensure the counters are not negative
counters.uploads = Math.max(counters.uploads, 0);
counters.errors = Math.max(counters.errors, 0);
```

There may also be a way to achieve the aims of this function (and enforce its minima) more elegantly via model relays and transform functions.